src/Features/CSharp/Portable/Completion/CompletionProviders/FunctionPointerUnmanagedCallingConventionCompletionProvider.cs

```diff
@@ -0,0 +1,139 @@
+        {
+            try
+            {
+                var document = context.Document;
+                var position = context.Position;
+                var cancellationToken = context.CancellationToken;
+
+                var syntaxTree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                if (syntaxTree.IsInNonUserCode(position, cancellationToken))
+                {
+                    return;
+                }
+
+                var token = syntaxTree
+                    .FindTokenOnLeftOfPosition(position, cancellationToken)
+                    .GetPreviousTokenIfTouchingWord(position);
+
+                if (!token.IsKind(SyntaxKind.OpenBracketToken, SyntaxKind.CommaToken))
+                {
+                    return;
+                }
+
+                if (token.Parent is not FunctionPointerUnmanagedCallingConventionListSyntax callingConventionList)
+                {
+                    return;
+                }
+
+                var contextPosition = token.SpanStart;
+                var semanticModel = await document.ReuseExistingSpeculativeModelAsync(callingConventionList, cancellationToken).ConfigureAwait(false);
+
+                var completionItems = new HashSet<CompletionItem>(CompletionItemComparer.Instance);
+                AddTypes(completionItems, contextPosition, semanticModel, cancellationToken);
+
+                // Even if we didn't have types, there are four magic calling conventions recognized regardless.
+                // We add these after doing the type lookup so if we had types we can show that instead
+                foreach (var callingConvention in s_predefinedCallingConventions)
+                {
+                    completionItems.Add(CompletionItem.Create(callingConvention, tags: GlyphTags.GetTags(Glyph.Keyword)));
+                }
+
+                context.AddItems(completionItems);
+            }
+            catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCanceled(e))
+            {
+                // nop
+            }
+        }
+
+        private static void AddTypes(HashSet<CompletionItem> completionItems, int contextPosition, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            // We have to find the set of types that meet the criteria listed in
+            // https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md#mapping-the-calling_convention_specifier-to-a-callkind
+            // We skip the check of an type being in the core assembly since that's not really necessary for our work.
+            var compilerServicesNamespace = semanticModel.Compilation.GlobalNamespace.GetQualifiedNamespace("System.Runtime.CompilerServices");
+            if (compilerServicesNamespace == null)
+            {
+                return;
+            }
+
+            foreach (var type in compilerServicesNamespace.GetTypeMembers())
```

Comment:
this only returns direct members right?  we don't need to filter out type members in child namespaces?

Reply:
Correct.